### PR TITLE
Masking legacy HB- energy correction for Run 3

### PR DESCRIPTION
**Pull request by igv4321**
#### PR description:

Adding an era-dependent mask so that legacy hardwired HB- energy corrections are not applied for Run 3. These corrections mitigate a hardware defect in the laser calibration system (in the light attenuation filter). They are applied only for data, as this defect is not modelled in the MC. The hardware will be fixed later this year, so these corrections are not needed in the future.

Also, very minor optimizations are made to the M0 rechit time code (no changes to the algorithm).

This is a pure maintenance PR. It is not supposed to modify any relval results.

#### PR validation:

"runTheMatrix" script was run and the results appear to be normal.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/HcalRecAlgos/src/parseHBHEPhase1AlgoDescription.cc [hunk 1]
```diff
@@ -125,9 +125,21 @@ parseHBHEPhase1AlgoDescription(const edm::ParameterSet& ps)
...
 
     return algo;
 }
+
+
+edm::ParameterSetDescription fillDescriptionForParseHBHEPhase1Algo()
+{
+    edm::ParameterSetDescription desc;
+
+    desc.setAllowAnything();
```
Comment (perrotta) [inline RecoLocalCalo/HcalRecAlgos/src/parseHBHEPhase1AlgoDescription.cc]: Why this ?
```
    desc.setAllowAnything();
```

**Comment by igv4321** [comment]
This is how it was originally in HBHEPhase1Reconstructor.cc. Taking into account the fact that the parameter sets can be different depending on the values of "useMahi", "useM2", and "useM3" (including/excluding various subalgorithms), creating a complete description taking into account all possible combinations of these settings would be complicated and error-prone.

**Comment by perrotta** [comment]
Thank you @igv4321 for your explanation, but I'm still trying to understand...

- The PSet "algorthm" can be different in the different reco algo (Mahi, M2, M3, M0...) and therefore you ```setAllowAnything()``` in that PSet to take possible different parameters into account for the different algos.
- You decide however to expicitly define "applyLegacyHBMCorrection"  in the ParameterSetDescription: https://github.com/cms-sw/cmssw/blob/e507dcf7a51ba2b2cd58f7005b4569794944a9aa/RecoLocalCalo/HcalRecAlgos/src/parseHBHEPhase1AlgoDescription.cc#L142
- That parameter is the only one that you define explicitly  in the ParameterSetDescription: why?
- There are indeed other parameters which are in common to all algos and could be explicitly defined there as well: "useM2", "useM3", "useMahi", "firstSampleShift", "samplesToAdd", "correctionPhaseNS", "tdcTimeShift", "correctForPhaseContainment" 
- Rephrasing my question in the light of what above: why only "applyLegacyHBMCorrection" is explicitly defined in the PSet, and not all other common parameters? IOW, what prevented you using the same ```add_param_set``` macro that you used before?

**Comment by igv4321** [comment]
Oh, the reason for singling out the "applyLegacyHBMCorrection" parameter has more to do with our practices than with the code design. This is a new parameter, and HBHEPhase1Reconstructor is a part of the HLT sequence. At some point in the past, HLT people elected to maintain their own set of complete CMSSW configurations, and there are many (inside HLTrigger/Configuration/python). I would either have to modify all of those configs or give the new parameter a default value in the descriptions -- otherwise HLT would not run in the relvals.

The new parameter will eventually percolate into the HLT configs, but for now it is easier to define it here.

**Comment by perrotta** [comment]
Ok, you are using the fillDescriptions method just as a replacement of the HLT customization, and not for its full potentiality. This still does make sense.

However, since you are already defining such a fillDescriptionForParseHBHEPhase1Algo() method, I would profit of it and just add also there all other parameters which are in common to all algos (the same I listed above, I'd say).  This could eventually become the basis for a future full transition to the usage of fillDesctiptions here.

**Comment by igv4321** [comment]
OK, added descriptions for other common parameters

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- The addition of an era-switch guarantees that  HB- energy corrections are not applied for Run 3
- Those corrections are only effective on real data, while they are not used in MC: therefore the switch doesn't affect simulated run3 data, while nothing changes for all other workflows
- Jenkins tests pass and show no differences

**Comment by fabiocos** [comment]
+1

**Pull request merged**